cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/ProvisionerStore.java

```diff
@@ -0,0 +1,91 @@
+ */
+
+package co.cask.cdap.internal.provision;
+
+import co.cask.cdap.api.Transactional;
+import co.cask.cdap.api.Transactionals;
+import co.cask.cdap.api.data.DatasetContext;
+import co.cask.cdap.api.dataset.DatasetManagementException;
+import co.cask.cdap.data.dataset.SystemDatasetInstantiator;
+import co.cask.cdap.data2.dataset2.DatasetFramework;
+import co.cask.cdap.data2.dataset2.MultiThreadDatasetCache;
+import co.cask.cdap.data2.transaction.TransactionSystemClientAdapter;
+import co.cask.cdap.data2.transaction.Transactions;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.cdap.proto.id.ProgramRunId;
+import com.google.inject.Inject;
+import org.apache.tephra.RetryStrategies;
+import org.apache.tephra.TransactionSystemClient;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/**
+ * Wrapper around {@link ProvisionerDataset} that runs every operation within a transaction.
+ */
+public class ProvisionerStore {
+  private final DatasetFramework datasetFramework;
+  private final Transactional transactional;
+
+  @Inject
+  ProvisionerStore(DatasetFramework datasetFramework, TransactionSystemClient txClient) {
+    this.datasetFramework = datasetFramework;
+    this.transactional = Transactions.createTransactionalWithRetry(
+      Transactions.createTransactional(new MultiThreadDatasetCache(new SystemDatasetInstantiator(datasetFramework),
+                                                                   new TransactionSystemClientAdapter(txClient),
+                                                                   NamespaceId.SYSTEM,
+                                                                   Collections.emptyMap(), null, null)),
+      RetryStrategies.retryOnConflict(20, 100)
+    );
+  }
+
+  public List<ClusterInfo> listClusterInfo() {
+    return Transactionals.execute(transactional, datasetContext -> {
+      ProvisionerDataset table = getTable(datasetContext);
+      return table.listClusterInfo();
+    });
+  }
+
+  @Nullable
+  public ClusterInfo getClusterInfo(ProgramRunId programRunId) {
+    return Transactionals.execute(transactional, datasetContext -> {
+      ProvisionerDataset table = getTable(datasetContext);
+      return table.getClusterInfo(programRunId);
+    });
+  }
+
+  public void putClusterInfo(ClusterInfo clusterInfo) {
+    Transactionals.execute(transactional, datasetContext -> {
+      ProvisionerDataset table = getTable(datasetContext);
+      table.putClusterInfo(clusterInfo);
+    });
+  }
+
+  public void deleteClusterInfo(ProgramRunId programRunId) {
+    Transactionals.execute(transactional, datasetContext -> {
+      ProvisionerDataset table = getTable(datasetContext);
+      table.deleteClusterInfo(programRunId);
+    });
+  }
+
+  private ProvisionerDataset getTable(DatasetContext context) throws IOException, DatasetManagementException {
+    ProvisionerDataset.createIfNotExists(datasetFramework);
```

Comment:
Better follow the https://github.com/caskdata/cdap/blob/develop/cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/LineageDataset.java#L114

so that it doesn't need to check with `DatasetFramework` for every call.